draft-ietf-mls-protocol.md

```diff
@@ -1594,13 +1594,34 @@ struct {
     uint16 count;
 } PSKLabel;
 
-psk_input_[i] = KDF.Extract(0, psk_[i])
-psk_secret_[i] = ExpandWithLabel(psk_input_[i], "derived psk", PSKLabel, KDF.Nh)
-psk_secret     = psk_secret_[0] || ... || psk_secret_[n-1]
+psk_extracted_[i] = KDF.Extract(0, psk_[i])
+psk_input_[i] = ExpandWithLabel(psk_extracted_[i], "derived psk", PSKLabel, KDF.Nh)
```

Comment:
```suggestion
psk_input_[i] = ExpandWithLabel(psk_extracted_[i], sprintf("derived psk %i", i), PSKLabel, KDF.Nh)
```
To err on the side of caution, I'd like to bake the structure of the cascade into the keys themselves. This ensures that you can't "extend" a cascade by taking the result of a cascade and use it to start a new one. Not something that would happen in the wild, I think, but it makes proofs easier and doesn't cost anything. The simple implementation you suggest would still work, you just have to add a counter.

(Not sure if the suggested line above is probably not what we want in the spec, but I think you understand what I mean.)

Reply:
Ah, you are right, of course.